cpp/ql/src/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -91,13 +91,30 @@ private float wideningUpperBounds(ArithmeticType t) {
   result = 1.0 / 0.0 // +Inf
 }
 
+/**
+ * Gets the value of the expression `e`, if it is a constant.
+ * This predicate also handles the case of constant variables initialized in compilation units,
+ * which doesn't necessarily have a getValue() result from the extractor.
+ */
+private string getValue(Expr e) {
+  if exists(e.getValue())
+  then result = e.getValue()
+  else
+    exists(VariableAccess access, Variable v |
+      e = access and
+      v = access.getTarget() and
+      v.getType().isConst() and
+      result = getValue(v.getAnAssignedValue())
```

Comment:
I don't think this value will always be unique. What if you have `const int C = 1;` in one file and `const int C = 2;` in another? And then `extern const int C;` in a third, which accesses `C`. Here I'm thinking C, not C++. IIUC, constants have external linkage in C and can be used like this, but I haven't checked.

You can use the trick of requiring that `result = min(...) and result = max(...)` to ensure there's one unique result. Or you can find lower and upper bound separately with `min` and `max` respectively.

Reply:
> IIUC, constants have external linkage in C and can be used like this, but I haven't checked.

I was not aware of this difference between C and C++! That should indeed be handled.

Reply:
I chose to keep the `getValue()` predicate as is (to avoid the need for `monotonicAggregates`), and instead do the `result = min(...) and result = max(...)` trick in a separate "top level" predicate. The tuple count looked fine on the projects I tested.

Reply:
I think I've given you bad advice here. It looks like the callers of `getValue` are able to tolerate multiple results. The only caller for which multiple values might matter is `getTruncatedLowerBounds` (and its twin for upper bounds). Most callers of `getTruncatedLowerBounds` are in the main recursion, where multiple results are tolerated as long as they are not numerous enough to kill performance. The value ends up in `lowerBound`, where a `min` is applied.

Reply:
I see. I've reverted this specific change now. Do you think we need to add a `rank[1..10]` expression (or something similar) to ensure that there aren't too many results?

Reply:
No, I think it's good now. If the results should explode on some particular codebase, then I'd like to understand why that happens before we choose a way to mitigate it.